geomstats/geometry/siegel.py

```diff
@@ -0,0 +1,614 @@
+        max_eigenvalues = gs.max(eigenvalues, axis=-1) ** 0.5
+
+        if ndim == 2:
+            if max_eigenvalues >= 1 - atol:
+                projected *= gs.cast((1 - atol) / max_eigenvalues, dtype=data_type)
+
+        elif ndim == 3:
+            for i_sample in range(n_samples):
+                if max_eigenvalues[i_sample] > 1 - atol:
+                    projected *= gs.cast(
+                        (1 - atol) / max_eigenvalues[i_sample], dtype=data_type
+                    )
+
+        return projected
+
+    def random_point(self, n_samples=1, bound=1.0):
+        """Generate random points in the Siegel space.
+
+        Parameters
+        ----------
+        n_samples : int
+            Number of samples.
+            Optional, default: 1.
+        bound : float
+            Bound of the interval in which to sample in the tangent space.
+            Optional, default: 1.
+
+        Returns
+        -------
+        samples : array-like, shape=[..., n, n]
+            Points sampled in the Siegel space.
+        """
+        n = self.n
+        size = (n_samples, n, n) if n_samples != 1 else (n, n)
+
+        samples = gs.random.rand(*size, dtype=gs.get_default_cdtype())
+        samples += 1j * gs.random.rand(*size, dtype=gs.get_default_cdtype())
+        samples -= 0.5 + 0.5j
+        samples /= 2 * n
+        samples *= 1 - gs.atol
+        samples *= bound
+        return samples
+
+    def random_tangent_vec(self, base_point=None, n_samples=1):
+        """Sample on the tangent space of Siegel space from the uniform distribution.
+
+        Parameters
+        ----------
+        n_samples : int
+            Number of samples.
+            Optional, default: 1.
+        base_point : array-like, shape=[..., n, n]
+            Base point of the tangent space.
+            Optional, default: None.
+
+        Returns
+        -------
+        samples : array-like, shape=[..., n, n]
+            Points sampled in the tangent space at base_point.
+        """
+        n = self.n
+        size = (n_samples, n, n) if n_samples != 1 else (n, n)
+
+        samples = gs.random.rand(*size, dtype=gs.get_default_cdtype())
+        samples += 1j * gs.random.rand(*size, dtype=gs.get_default_cdtype())
+        samples *= 2
+        samples -= 1 + 1j
+
+        return samples
+
+
+class SiegelMetric(ComplexRiemannianMetric):
+    """Class for the Siegel metric.
+
+    Parameters
+    ----------
+    n : int
+        Integer representing the shape of the matrices: n x n.
+    scale : float
+        Scale of the complex Poincare metric.
+        Optional, default: 1.
+    """
+
+    def __init__(self, n, scale=1.0, **kwargs):
+        dim = int(n**2)
+        super().__init__(
+            dim=dim,
+            shape=(n, n),
+            signature=(dim, 0),
+        )
+        self.n = n
+        self.scale = scale
+
+    def inner_product(self, tangent_vec_a, tangent_vec_b, base_point):
+        """Compute the Siegel inner-product.
```

Comment:
It is a long expression. It would be nice to have it here (specially for understanding the code).

Reply:
Done.